helix-term/src/ui/editor.rs

```diff
@@ -697,6 +738,242 @@ impl EditorView {
+        let syntax = doc.syntax()?;
+        let tree = syntax.tree();
+        let text = doc.text().slice(..);
+        let viewport = view.inner_area(doc);
+        let cursor_byte = doc.selection(view.id).primary().cursor(text);
+
+        // Use the cached nodes to determine the current topmost viewport
+        let anchor_line = text.char_to_line(view.offset.anchor);
+        let top_first_byte =
+            text.line_to_byte(anchor_line + nodes.as_deref().map_or(0, |v| v.len()));
+
+        let visual_cursor_pos = view
+            .screen_coords_at_pos(doc, text, cursor_byte)
+            .unwrap_or_default()
+            .row as u16;
+
+        if visual_cursor_pos == 0 {
+            return None;
+        }
+
+        if let Some(nodes) = nodes {
+            if nodes.iter().any(|node| {
+                node.top_first_byte == top_first_byte && (visual_cursor_pos as usize) >= nodes.len()
+            }) {
+                return Some(nodes.to_vec());
+            }
+        }
+
+        let context_nodes = doc
+            .language_config()
+            .and_then(|lang| lang.context_query())?;
+
+        let mut cursor_parent = tree
+            .root_node()
+            .descendant_for_byte_range(cursor_byte, cursor_byte)
+            .and_then(|n| n.parent());
+
+        let mut cursor_nodes = Vec::new();
+        while let Some(node) = cursor_parent {
+            cursor_nodes.push(node);
+            cursor_parent = node.parent();
+        }
+
+        let mut parent = tree
+            .root_node()
+            .descendant_for_byte_range(top_first_byte, top_first_byte)
+            .and_then(|n| n.parent());
+
+        // context is list of numbers of lines that should be rendered in the LSP context
+        let mut context: Vec<StickyNode> = Vec::new();
+
+        while let Some(node) = parent {
+            // skip all the nodes that aren't in the scope of the cursor
+            if !cursor_nodes
+                .iter()
+                .any(|cursor_node| cursor_node.start_byte() == node.start_byte())
+            {
+                parent = node.parent();
+                continue;
+            }
+
+            let line = text.byte_to_line(node.start_byte());
+
+            // if parent of previous node is still on the same line, use the parent node
+            if let Some(prev_line) = context.last() {
+                if prev_line.line_nr == line {
+                    context.pop();
+                }
+            }
+
+            let mut cursor = QueryCursor::new();
+            let query = &context_nodes.query;
+            let query_nodes = cursor.matches(query, node, RopeProvider(text));
+
+            let query_ranges = query_nodes
+                .flat_map(|qnode| {
+                    qnode
+                        .captures
+                        .iter()
+                        .map(|capture| capture.node.byte_range())
+                })
+                .collect::<Vec<std::ops::Range<usize>>>();
```

Comment:
You're allocationg a vec only to immediately iterate on it, it should be possible to avoid the allocation entirely